components/inputs/input-date-time.js

```diff
@@ -138,30 +157,98 @@ class InputDateTime extends LocalizeCoreElement(RtlMixin(LitElement)) {
 		if (elem) elem.focus();
 	}
 
+	get validationMessageRangeOverflow() {
+		let failureText = '';
+		if (this.minValue && this.maxValue) {
+			failureText = this.localize(
+				`${this._namespace}.errorOutsideRange`, {
+					minDate: formatDateTime(getDateFromISODateTime(this.minValue), {format: 'medium'}),
+					maxDate: formatDateTime(getDateFromISODateTime(this.maxValue), {format: 'medium'})
+				}
+			);
+		} else if (this.maxValue) {
+			failureText = this.localize(
+				`${this._namespace}.errorMaxDateOnly`, {
+					maxDate: formatDateTime(getDateFromISODateTime(this.maxValue), {format: 'medium'})
+				}
+			);
+		}
+		return failureText;
+	}
+
+	get validationMessageRangeUnderflow() {
+		let failureText = '';
+		if (this.minValue && this.maxValue) {
+			failureText = this.localize(
+				`${this._namespace}.errorOutsideRange`, {
+					minDate: formatDateTime(getDateFromISODateTime(this.minValue), {format: 'medium'}),
+					maxDate: formatDateTime(getDateFromISODateTime(this.maxValue), {format: 'medium'})
+				}
+			);
+		} else if (this.minValue) {
+			failureText = this.localize(
+				`${this._namespace}.errorMinDateOnly`, {
+					minDate: formatDateTime(getDateFromISODateTime(this.minValue), {format: 'medium'})
+				}
+			);
+		}
+		return failureText;
+	}
+
 	_dispatchChangeEvent() {
 		this.dispatchEvent(new CustomEvent(
 			'change',
 			{ bubbles: true, composed: false }
 		));
 	}
 
-	_handleDateChange(e) {
-		// TODO: handle validation with min and max value
+	async _handleDateChange(e) {
 		const newDate = e.target.value;
 		if (!newDate) {
 			this.value = '';
 		} else {
 			const time = this.shadowRoot.querySelector('d2l-input-time').value;
 			this.value = getUTCDateTimeFromLocalDateTime(newDate, time);
 		}
+		await this._validate();
 		this._dispatchChangeEvent();
 	}
 
-	_handleTimeChange(e) {
-		// TODO: handle validation with min and max value
+	async _handleDropdownToggle(e) {
+		this._dropdownOpened = e.detail.opened;
+		if (e.target.tagName === 'D2L-INPUT-TIME' && !this._dropdownOpened) {
+			await this.updateComplete;
+			this._handleInputTimeFocus();
+		}
+	}
+
+	_handleInputTimeBlur() {
+		const tooltip = this.shadowRoot.querySelector('d2l-tooltip');
+		if (tooltip) tooltip.hide();
+	}
+
+	_handleInputTimeFocus() {
+		const tooltip = this.shadowRoot.querySelector('d2l-tooltip');
+		if (tooltip) tooltip.show();
+	}
```

Comment:
This feels like something that could be made a lot easier. 🤔

Not sure if anyone else has any thoughts but could it be worth expanding tooltip to support multiple targets? We could then do `for="${dateId} ${timeId}"` and the tooltip would position itself relative to the smallest rectangle that contains both along with handling focus events from both?

It would be a bit of work to add so probably not immediately but maybe something that'd be of value?

Reply:
I like that idea!